R/detect_mistakes.R

```diff
@@ -1,281 +1,302 @@
-  # else it's wrong
+  # 1. If the code contains a bare value, then the user and solution value
+  # should be identical.
+  # BUT WHAT IF ONE IS A CALL THAT EVALUATES TO THE VALUE OF THE OTHER?
   if (!is.call(user) || !is.call(solution)) {
     if (!identical(user, solution)) {
-      return(wrong_value(this = deparse_to_string(user),
-                  that = solution)
+      return(
+        wrong_value(
+          this = submitted,
+          that = solution,
+          this_name = enclosing_arg,
+          enclosing_call = enclosing_call
+        )
       )
     }
   }
-
-  user_len <- length(user)
-  solution_len <- length(solution)
+  # We can assume anything below here is a call
 
   user_names <- rlang::names2(user)
   solution_names <- rlang::names2(solution)
-
-  if (user_len > solution_len) {
-    for (i in seq_len(user_len)) {
-      if (i > solution_len || user_names[i] != solution_names[i]) {
-        return(
-          surplus_argument(
-            this_call = user,
-            this = user[[i]],
-            this_name = user_names[[i]]
-          )
+  
+  # Dividing cases into groups based on the relative lengths of the user's code
+  # and the solution code produces unitelligible messages as in issue #84. To
+  # produce more transparent messages that accord with how a user thinks of calls,
+  # check these things in this order:
+  
+  # 2. Check that the user and the solution use the same call
+  # SHOULD WE HAVE A TARGETED WRONG CALL FUNCTION?
+  if (!identical(user[[1]], solution[[1]])) {
+    return(
+      wrong_call(
+        this = user,
+        that = solution,
+        this_name = enclosing_arg,
+        enclosing_call = enclosing_call
+      )
+    )
+  }
+  
+  # 3. Check that the user code is not malformed and can be safely passed to
+  # call_standardise_formals(), which uses match.call(). Malformed code may
+  # contain an unused argument, may contain multiple arguments whose names
+  # partially match the same formal, or an argument whose name partially matches
+  # more than one formal.
+  user_args <- as.list(user)
+  real_user_names <- user_names[user_names != ""]
+  
+  solution_args <- as.list(solution)
+  real_solution_names <- solution_names[solution_names != ""]
+  
+  ## Remove exact matches from further scrutiny
+  for (name in real_user_names) {
+    if (name %in% real_solution_names) {
+      user_args[[name]] <- NULL
+      solution_args[[name]] <- NULL
+    }
+  }
+  
+  ## Check remaining arguments for partial matches
+  user_names <- rlang::names2(user_args)
+  remaining_user_names <- user_names[user_names != ""]
+  
+  if (length(remaining_user_names) > 0) {
+    solution_names <- rlang::names2(solution_args)
+    remaining_solution_names <- solution_names[solution_names != ""]
+    
+    ## Do any solution names partially match multiple user names?
+    pmatches_per_formal <- function(solution_name) {
+      sum(startsWith(solution_name, remaining_user_names))
+    }
```

Comment:
Check out `match.arg`

Reply:
`match.arg` won't work here because it throws an error if the solution name is not in the `remaining_user_names`